[R3 Router Replay] The seq_len of routed_experts does not match encoded length; expected same length or one less.

### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new question or discussion topic. / 我已经搜索过现有的 issues，确认这是一个新的问题与讨论。

### Question Description / 问题描述

I encountered the following error when doing GSPO training (with R3 router replay enabled) on Qwen3.5-35B-A3B:

```bash
[rank4]: Traceback (most recent call last):
[rank4]:   File "/mnt/ms-swift/swift/cli/_megatron/rlhf.py", line 24, in <module>
[rank4]:     megatron_rlhf_main()
[rank4]:   File "/mnt/ms-swift/swift/megatron/pipelines/train/rlhf.py", line 73, in megatron_rlhf_main
[rank4]:     return MegatronRLHF(args).main()
[rank4]:            ^^^^^^^^^^^^^^^^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/pipelines/base.py", line 52, in main
[rank4]:     result = self.run()
[rank4]:              ^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/megatron/pipelines/train/sft.py", line 72, in run
[rank4]:     trainer.train(train_dataset, val_dataset)
[rank4]:   File "/mnt/ms-swift/swift/megatron/trainers/grpo_trainer.py", line 73, in train
[rank4]:     super().train(train_dataset, val_dataset)
[rank4]:   File "/mnt/ms-swift/swift/megatron/trainers/base.py", line 707, in train
[rank4]:     self.run_train_step(train_data_iterator, val_data_iterator)
[rank4]:   File "/mnt/ms-swift/swift/megatron/trainers/base.py", line 647, in run_train_step
[rank4]:     metrics, grad_norm, update_successful = self.train_step(train_data_iterator)
[rank4]:                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/megatron/trainers/base.py", line 890, in train_step
[rank4]:     data_iterator = self._replace_data_iterator(train_data_iterator)
[rank4]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/megatron/trainers/grpo_trainer.py", line 258, in _replace_data_iterator
[rank4]:     micro_batch_data = self._generate_and_score_completions(rollout_batch)
[rank4]:                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/megatron/trainers/grpo_trainer.py", line 314, in _generate_and_score_completions
[rank4]:     model_inputs, grpo_batch = collate_to_grpo_micro_batch(
[rank4]:                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/rlhf_trainers/utils.py", line 2133, in collate_to_grpo_micro_batch
[rank4]:     routed_experts = build_routed_experts_batch(
[rank4]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank4]:   File "/mnt/ms-swift/swift/rlhf_trainers/utils.py", line 2072, in build_routed_experts_batch
[rank4]:     raise AssertionError(f'The seq_len of routed_experts({experts_seq_len}) does not match encoded length '
[rank4]: AssertionError: The seq_len of routed_experts(1854) does not match encoded length (1856); expected same length or one less.
```

Here is the related parameter settings in my training scripts:
```bash
megatron rlhf \
  --rlhf_type grpo \
  ... \
  --enable_thinking false \
  --context_parallel_size 4 \
  --tensor_model_parallel_size 1 \
  --expert_model_parallel_size 8 \
  --pipeline_model_parallel_size 1 \
  --moe_permute_fusion true \
  --moe_expert_capacity_factor 2 \
  --micro_batch_size 1 \
  --global_batch_size 512 \
  --steps_per_generation 1 \
  --num_generations 8 \
  --use_vllm true \
  --vllm_mode colocate \
  --vllm_gpu_memory_utilization 0.45 \
  --vllm_tensor_parallel_size 8 \
  --vllm_max_model_len 36000 \
  --vllm_mm_processor_cache_gb 0 \
  --max_length 32768 \
  --tuner_type full \
  ... \
  --bf16 true \
  --beta 0.04 \
  --importance_sampling_level sequence \
  --router_replay_mode R3 \
  ... \
  --loss_type grpo \
  ...
```
And my related package versions:
```
vllm: 0.23.1.dev0+g0fc695fc6.d20260703.cu128
torch: 2.11.0+cu128
transformers: 5.12.1
ms_swift: 4.5.0.dev0
```

--- a/swift/rlhf_trainers/utils.py
+++ b/swift/rlhf_trainers/utils.py
@@ -2047,13 +2047,6 @@ def build_routed_experts_batch(
     Shared by all backends. Each ``sample`` is an :class:`OnPolicySample` carrying
     ``routed_experts`` (per-sample, seq-first) and ``encoded['length']``. Returns
     ``None`` when no sample provides routing (and mode is not ``R3``).
-
-    ``expected_lens`` carries the per-sample encode lengths captured *before*
-    ``template.data_collator`` ran. This matters under Megatron CP:
-    ``_handle_megatron_cp`` mutates each ``encoded`` in place, rounding
-    ``encoded['length']`` up to a multiple of ``cp_size * 2``, so reading it here
-    would compare vLLM's unpadded routing against a padded length. When omitted,
-    it falls back to ``encoded['length']``.
     """
     if not samples or all(getattr(s, 'routed_experts', None) is None for s in samples):
         return None

--- a/swift/rlhf_trainers/utils.py
+++ b/swift/rlhf_trainers/utils.py
@@ -2040,12 +2040,20 @@ def build_routed_experts_batch(
     template: Template,
     device: torch.device,
     router_replay_mode: str = 'disabled',
+    expected_lens: Optional[List[Optional[int]]] = None,
 ) -> Optional[torch.Tensor]:
     """Build the batched ``routed_experts`` model input from per-sample R3 routing.
 
     Shared by all backends. Each ``sample`` is an :class:`OnPolicySample` carrying
     ``routed_experts`` (per-sample, seq-first) and ``encoded['length']``. Returns
     ``None`` when no sample provides routing (and mode is not ``R3``).
+
+    ``expected_lens`` carries the per-sample encode lengths captured *before*
+    ``template.data_collator`` ran. This matters under Megatron CP:
+    ``_handle_megatron_cp`` mutates each ``encoded`` in place, rounding
+    ``encoded['length']`` up to a multiple of ``cp_size * 2``, so reading it here
+    would compare vLLM's unpadded routing against a padded length. When omitted,
+    it falls back to ``encoded['length']``.
     """
     if not samples or all(getattr(s, 'routed_experts', None) is None for s in samples):
         return None
@@ -2058,16 +2066,20 @@ def build_routed_experts_batch(
         current_seq_lengths[n_samples - 1] = seq_lengths[n_samples - 1:].sum()
 
     routed_tensors: List[torch.Tensor] = []
-    for sample, cur_seq_len in zip(samples, current_seq_lengths):
+    for i, (sample, cur_seq_len) in enumerate(zip(samples, current_seq_lengths)):
         routed_value = getattr(sample, 'routed_experts', None)
         if routed_value is None:
             if router_replay_mode == 'R3':
                 raise AssertionError('When router_replay_mode = R3, routed_experts must be in rollout data')
             return None
         routed = _normalize_routed_experts_tensor(routed_value)
-        expected_len = (sample.encoded or {}).get('length')
+        if expected_lens is not None:
+            expected_len = expected_lens[i]
+        else:
+            expected_len = sample.encoded.get('length')
         experts_seq_len = int(routed.shape[0])
         if router_replay_mode == 'R3' and expected_len is not None:
+            # vLLM may omit the final token's routing, hence ``expected_len - 1``.
             if experts_seq_len not in (expected_len, expected_len - 1):
                 raise AssertionError(f'The seq_len of routed_experts({experts_seq_len}) does not match encoded length '
                                      f'({expected_len}); expected same length or one less.')
@@ -2112,6 +2124,9 @@ def collate_to_grpo_micro_batch(
     distributed communication happens here.
     """
     encoded_list = [s.encoded for s in samples]
+    # Snapshot encode lengths before ``data_collator``: under Megatron CP it mutates each
+    # ``encoded`` in place, rounding ``length`` up to a multiple of ``cp_size * 2``.
+    expected_lens = [e.get('length') for e in encoded_list]
     model_inputs = to_device(template.data_collator(encoded_list, padding_to=padding_to), device)
 
     labels = model_inputs['labels']
@@ -2137,6 +2152,7 @@ def collate_to_grpo_micro_batch(
         template=template,
         device=device,
         router_replay_mode=router_replay_mode,
+        expected_lens=expected_lens,
     )
     if routed_experts is not None:
         model_inputs['routed_experts'] = routed_experts

